In [1]:
import pandas as pd

Problem Statement: https://www.kaggle.com/c/house-prices-advanced-regression-techniques
Dataset: https://github.com/abulbasar/data/tree/master/kaggle-houseprice

In [2]:
df = pd.read_csv("/data/kaggle/data_combined_cleaned.csv")
df = df.drop(columns=["Id"])
df = df[~df.SalesPrice.isnull()]
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave  None      Reg   
1          20       RL         80.0     9600   Pave  None      Reg   
2          60       RL         68.0    11250   Pave  None      IR1   
3          70       RL         60.0     9550   Pave  None      IR1   
4          60       RL         84.0    14260   Pave  None      IR1   

  LandContour LotConfig LandSlope    ...     PoolArea PoolQC Fence  \
0         Lvl    Inside       Gtl    ...            0   None  None   
1         Lvl       FR2       Gtl    ...            0   None  None   
2         Lvl    Inside       Gtl    ...            0   None  None   
3         Lvl    Corner       Gtl    ...            0   None  None   
4         Lvl       FR2       Gtl    ...            0   None  None   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition SalesPrice  
0        None       0       2    2008        WD         Normal   208500.0  
1        None       0       5    2007        WD         Normal   181500.0  
2        None       0       9    2008        WD         Normal   223500.0  
3        None       0       2    2006        WD        Abnorml   140000.0  
4        None       0      12    2008        WD         Normal   250000.0  

[5 rows x 79 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 0 to 1459
Data columns (total 79 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1460 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            1460 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non-null object
MasVnrType       1460 no

# Exercise:
- Build a model to predict the SalesPrice using other columns as features. Split the dataset (1460 records) into training and test sets with 70/30 ratio. Take log of SalesPrice for prediction to reduce the impact of the outliers.
- What is the R2 score and rmse score? 
- Fit Lasso model. Test a few alpha values (>0) and see which gives a good r2 score. 

In [4]:
import numpy as np
from sklearn import preprocessing, model_selection, linear_model, metrics, pipeline

In [5]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.LinearRegression())
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]
pd.DataFrame({"feature": columns, "coefficient": est.coef_})

training r2 0.9494518297088844
test r2 -3.0715943247509893e+22
training rmse: 0.08606817747710854
test rmse: 76190488351.17424


feature   coefficient
0               MSSubClass  1.248159e-02
1              LotFrontage  1.963124e-03
2                  LotArea  3.151273e-02
3              OverallQual  5.675519e-02
4              OverallCond  3.468973e-02
5                YearBuilt  5.024337e-02
6             YearRemodAdd  1.540301e-02
7               MasVnrArea -3.995381e-03
8               BsmtFinSF1  4.530257e+10
9               BsmtFinSF2  1.612199e+10
10               BsmtUnfSF  4.288808e+10
11             TotalBsmtSF -4.382796e+10
12                1stFlrSF -1.112376e+11
13                2ndFlrSF -1.233520e+11
14            LowQualFinSF -1.362186e+10
15               GrLivArea  1.501695e+11
16            BsmtFullBath  8.699417e-03
17            BsmtHalfBath  1.619369e-03
18                FullBath  7.283211e-03
19                HalfBath  9.532332e-03
20            BedroomAbvGr  1.488495e-02
21            KitchenAbvGr -1.001364e-02
22            TotRmsAbvGrd -1.378536e-03
23              Fireplaces  1.032925e-02
24             GarageYrBlt -1.290452e-02
25              GarageCars  5.177021e-03
26              GarageArea  3.925258e-02
27              WoodDeckSF  1.009178e-02
28             OpenPorchSF  9.440899e-03
29           EnclosedPorch  1.197055e-02
..                     ...           ...
228          GarageCond_Gd  2.391434e-02
229        GarageCond_None  7.930320e+10
230          GarageCond_Po  3.245735e-02
231          GarageCond_TA  6.799388e-02
232           PavedDrive_P -3.139496e-03
233           PavedDrive_Y  2.371550e-03
234              PoolQC_Fa  1.114522e+03
235              PoolQC_Gd  1.083305e-02
236            PoolQC_None  3.260802e-02
237             Fence_GdWo -1.957178e-03
238            Fence_MnPrv  6.403923e-04
239             Fence_MnWw  2.841771e-03
240             Fence_None  5.779266e-04
241       MiscFeature_None  1.070099e-01
242       MiscFeature_Othr  2.083778e-02
243       MiscFeature_Shed  9.071565e-02
244       MiscFeature_TenC  3.388132e-21
245           SaleType_CWD  9.831190e-04
246           SaleType_Con  5.665779e-03
247         SaleType_ConLD  7.680893e-03
248         SaleType_ConLI -4.849315e-03
249         SaleType_ConLw -3.542900e-04
250           SaleType_New  4.478741e-02
251           SaleType_Oth  8.874655e-03
252            SaleType_WD -5.639076e-03
253  SaleCondition_AdjLand  1.772881e-03
254   SaleCondition_Alloca  1.122475e-02
255   SaleCondition_Family  9.747744e-04
256   SaleCondition_Normal  2.903473e-02
257  SaleCondition_Partial -1.682472e-02

[258 rows x 2 columns]

In [6]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Lasso(alpha=0.01, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.8940480673803641
test r2 0.8785389987905549
training rmse: 0.12460754170603186
test rmse: 0.15150859029885974


feature  coefficient  coef_mag
3             OverallQual     0.104135  0.104135
15              GrLivArea     0.095563  0.095563
25             GarageCars     0.055896  0.055896
5               YearBuilt     0.029465  0.029465
39            MSZoning_RM    -0.026432  0.026432
4             OverallCond     0.023309  0.023309
6            YearRemodAdd     0.021713  0.021713
70   Neighborhood_NridgHt     0.020148  0.020148
60   Neighborhood_Crawfor     0.019489  0.019489
16           BsmtFullBath     0.018030  0.018030
90        Condition2_PosN    -0.015700  0.015700
0              MSSubClass    -0.014961  0.014961
178      BsmtFinType1_Unf    -0.014265  0.014265
210      FireplaceQu_None    -0.013760  0.013760
157      Foundation_PConc     0.012625  0.012625
18               FullBath     0.011998  0.011998
2                 LotArea     0.011859  0.011859
76   Neighborhood_StoneBr     0.011295  0.011295
22           TotRmsAbvGrd     0.011208  0.011208
80        Condition1_Norm     0.010955  0.010955
27             WoodDeckSF     0.010850  0.010850
235             PoolQC_Gd    -0.010616  0.010616
169       BsmtExposure_Gd     0.010399  0.010399
202       Functional_Maj2    -0.009603  0.009603
213     GarageType_Attchd     0.009391  0.009391
194          CentralAir_Y     0.009121  0.009121
119   Exterior1st_BrkFace     0.009079  0.009079
163         BsmtQual_None    -0.008641  0.008641
23             Fireplaces     0.008159  0.008159
250          SaleType_New     0.007931  0.007931
172     BsmtExposure_None    -0.007832  0.007832
12               1stFlrSF     0.007722  0.007722
118   Exterior1st_BrkComm    -0.007699  0.007699
49      LotConfig_CulDSac     0.007518  0.007518
61   Neighborhood_Edwards    -0.006650  0.006650
31            ScreenPorch     0.005994  0.005994
171       BsmtExposure_No    -0.005965  0.005965
151          ExterQual_TA    -0.005217  0.005217
152          ExterCond_Fa    -0.004778  0.004778
11            TotalBsmtSF     0.004723  0.004723
96         BldgType_Twnhs    -0.004638  0.004638
28            OpenPorchSF     0.004604  0.004604
207        Functional_Typ     0.004278  0.004278
201        KitchenQual_TA    -0.004073  0.004073
193          HeatingQC_TA    -0.003568  0.003568
190          HeatingQC_Fa    -0.003450  0.003450
58   Neighborhood_ClearCr     0.002836  0.002836
206        Functional_Sev    -0.002447  0.002447
127    Exterior1st_Stucco    -0.002415  0.002415
21           KitchenAbvGr    -0.002275  0.002275
116      RoofMatl_WdShngl     0.001826  0.001826
174      BsmtFinType1_GLQ     0.001595  0.001595
45           LotShape_Reg    -0.001593  0.001593
142    Exterior2nd_Stucco    -0.001218  0.001218
105       RoofStyle_Gable    -0.001104  0.001104
69   Neighborhood_NoRidge     0.000791  0.000791
44           LotShape_IR3    -0.000444  0.000444
64   Neighborhood_MeadowV    -0.000369  0.000369
255  SaleCondition_Family    -0.000161  0.000161

In [7]:
summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

feature  coefficient  coef_mag
3             OverallQual     0.104135  0.104135
15              GrLivArea     0.095563  0.095563
25             GarageCars     0.055896  0.055896
5               YearBuilt     0.029465  0.029465
39            MSZoning_RM    -0.026432  0.026432
4             OverallCond     0.023309  0.023309
6            YearRemodAdd     0.021713  0.021713
70   Neighborhood_NridgHt     0.020148  0.020148
60   Neighborhood_Crawfor     0.019489  0.019489
16           BsmtFullBath     0.018030  0.018030
90        Condition2_PosN    -0.015700  0.015700
0              MSSubClass    -0.014961  0.014961
178      BsmtFinType1_Unf    -0.014265  0.014265
210      FireplaceQu_None    -0.013760  0.013760
157      Foundation_PConc     0.012625  0.012625
18               FullBath     0.011998  0.011998
2                 LotArea     0.011859  0.011859
76   Neighborhood_StoneBr     0.011295  0.011295
22           TotRmsAbvGrd     0.011208  0.011208
80        Condition1_Norm     0.010955  0.010955
27             WoodDeckSF     0.010850  0.010850
235             PoolQC_Gd    -0.010616  0.010616
169       BsmtExposure_Gd     0.010399  0.010399
202       Functional_Maj2    -0.009603  0.009603
213     GarageType_Attchd     0.009391  0.009391
194          CentralAir_Y     0.009121  0.009121
119   Exterior1st_BrkFace     0.009079  0.009079
163         BsmtQual_None    -0.008641  0.008641
23             Fireplaces     0.008159  0.008159
250          SaleType_New     0.007931  0.007931
172     BsmtExposure_None    -0.007832  0.007832
12               1stFlrSF     0.007722  0.007722
118   Exterior1st_BrkComm    -0.007699  0.007699
49      LotConfig_CulDSac     0.007518  0.007518
61   Neighborhood_Edwards    -0.006650  0.006650
31            ScreenPorch     0.005994  0.005994
171       BsmtExposure_No    -0.005965  0.005965
151          ExterQual_TA    -0.005217  0.005217
152          ExterCond_Fa    -0.004778  0.004778
11            TotalBsmtSF     0.004723  0.004723
96         BldgType_Twnhs    -0.004638  0.004638
28            OpenPorchSF     0.004604  0.004604
207        Functional_Typ     0.004278  0.004278
201        KitchenQual_TA    -0.004073  0.004073
193          HeatingQC_TA    -0.003568  0.003568
190          HeatingQC_Fa    -0.003450  0.003450
58   Neighborhood_ClearCr     0.002836  0.002836
206        Functional_Sev    -0.002447  0.002447
127    Exterior1st_Stucco    -0.002415  0.002415
21           KitchenAbvGr    -0.002275  0.002275
116      RoofMatl_WdShngl     0.001826  0.001826
174      BsmtFinType1_GLQ     0.001595  0.001595
45           LotShape_Reg    -0.001593  0.001593
142    Exterior2nd_Stucco    -0.001218  0.001218
105       RoofStyle_Gable    -0.001104  0.001104
69   Neighborhood_NoRidge     0.000791  0.000791
44           LotShape_IR3    -0.000444  0.000444
64   Neighborhood_MeadowV    -0.000369  0.000369
255  SaleCondition_Family    -0.000161  0.000161

In [22]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 147595)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Ridge(alpha=40, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.9468756989746974
test r2 0.6160944791413361
training rmse: 0.09374379292490413
test rmse: 0.23569702086858121


feature  coefficient  coef_mag
235             PoolQC_Gd    -0.074129  0.074129
15              GrLivArea     0.062381  0.062381
3             OverallQual     0.053154  0.053154
38            MSZoning_RL     0.050967  0.050967
90        Condition2_PosN    -0.046524  0.046524
12               1stFlrSF     0.045143  0.045143
4             OverallCond     0.039686  0.039686
39            MSZoning_RM     0.039352  0.039352
11            TotalBsmtSF     0.036887  0.036887
13               2ndFlrSF     0.036532  0.036532
2                 LotArea     0.030100  0.030100
5               YearBuilt     0.029605  0.029605
36            MSZoning_FV     0.028246  0.028246
8              BsmtFinSF1     0.026761  0.026761
194          CentralAir_Y     0.025962  0.025962
60   Neighborhood_Crawfor     0.023025  0.023025
26             GarageArea     0.022416  0.022416
80        Condition1_Norm     0.021935  0.021935
164           BsmtQual_TA    -0.020234  0.020234
200        KitchenQual_Gd    -0.018930  0.018930
201        KitchenQual_TA    -0.018745  0.018745
256  SaleCondition_Normal     0.018529  0.018529
64   Neighborhood_MeadowV    -0.018049  0.018049
63    Neighborhood_IDOTRR    -0.017500  0.017500
202       Functional_Maj2    -0.017043  0.017043
70   Neighborhood_NridgHt     0.016382  0.016382
19               HalfBath     0.016008  0.016008
54          LandSlope_Sev    -0.015950  0.015950
22           TotRmsAbvGrd     0.015822  0.015822
207        Functional_Typ     0.015658  0.015658
..                    ...          ...       ...
143   Exterior2nd_VinylSd    -0.000991  0.000991
188          Heating_OthW     0.000909  0.000909
220      GarageFinish_RFn     0.000895  0.000895
87       Condition2_Feedr     0.000887  0.000887
212        FireplaceQu_TA    -0.000874  0.000874
147       MasVnrType_None     0.000800  0.000800
101     HouseStyle_2.5Unf    -0.000779  0.000779
77    Neighborhood_Timber    -0.000699  0.000699
20           BedroomAbvGr    -0.000698  0.000698
57   Neighborhood_BrkSide     0.000690  0.000690
137   Exterior2nd_ImStucc     0.000649  0.000649
138   Exterior2nd_MetalSd    -0.000649  0.000649
195      Electrical_FuseF     0.000585  0.000585
7              MasVnrArea     0.000577  0.000577
251          SaleType_Oth     0.000553  0.000553
94        BldgType_2fmCon     0.000531  0.000531
181      BsmtFinType2_LwQ     0.000522  0.000522
126     Exterior1st_Stone     0.000503  0.000503
72     Neighborhood_SWISU     0.000476  0.000476
122   Exterior1st_HdBoard     0.000456  0.000456
118   Exterior1st_BrkComm     0.000435  0.000435
141     Exterior2nd_Stone    -0.000421  0.000421
243      MiscFeature_Shed    -0.000417  0.000417
34                 MoSold     0.000394  0.000394
74   Neighborhood_SawyerW    -0.000363  0.000363
255  SaleCondition_Family    -0.000296  0.000296
205        Functional_Mod    -0.000287  0.000287
85        Condition1_RRNe    -0.000180  0.000180
44           LotShape_IR3     0.000067  0.000067
192          HeatingQC_Po     0.000003  0.000003

[251 rows x 3 columns]

In [9]:
from sklearn import feature_selection

In [18]:
scaler = preprocessing.StandardScaler()
X_train_std = scaler.fit_transform(X_train)
_, pvals = feature_selection.f_regression(X_train_std, y_train)
summary = pd.DataFrame({"feature": columns, "pval": pvals})
summary = summary.sort_values("pval")
summary = summary[summary.pval < 0.01] 
# pval < 0.01 are considered significant
# closer to 0 is more significant features
summary.iloc[:10,:]

/anaconda3/lib/python3.6/site-packages/sklearn/feature_selection/univariate_selection.py:299: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


feature           pval
3     OverallQual  1.543610e-232
15      GrLivArea  1.502663e-147
25     GarageCars  8.980028e-140
26     GarageArea  9.673012e-121
18       FullBath  7.435435e-106
11    TotalBsmtSF   4.707153e-98
151  ExterQual_TA   1.996594e-96
12       1stFlrSF   1.957045e-95
5       YearBuilt   2.946855e-85
6    YearRemodAdd   3.840149e-80

In [21]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Lasso(alpha=0.07, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.7395929485859631
test r2 0.746971064732701
training rmse: 0.19535127802592958
test rmse: 0.2186774107558645


feature  coefficient  coef_mag
3    OverallQual     0.140629  0.140629
15     GrLivArea     0.069171  0.069171
25    GarageCars     0.055262  0.055262
11   TotalBsmtSF     0.014343  0.014343
5      YearBuilt     0.012730  0.012730
6   YearRemodAdd     0.009382  0.009382
12      1stFlrSF     0.006673  0.006673
39   MSZoning_RM    -0.005534  0.005534
23    Fireplaces     0.000774  0.000774

In [25]:
pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Ridge(alpha=40, random_state=1))
])

scores = model_selection.cross_val_score(pipe, X=X_train, y=y_train, cv=5)
scores

array([0.89012794, 0.70703003, 0.81430385, 0.83800989, 0.87462756])

In [26]:
np.mean(scores)

0.8248198555818318

In [27]:
pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Ridge(alpha=40, random_state=1))
])

scores = model_selection.cross_val_score(pipe, X=X_train, y=y_train, cv=5)
np.mean(scores)

0.8248198555818318

In [28]:
pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Lasso(alpha=0.07, random_state=1))
])

scores = model_selection.cross_val_score(pipe, X=X_train, y=y_train, cv=5)
np.mean(scores)

0.746624200157168

In [29]:
# Hyper parameter tuning

In [41]:
np.linspace(0, 1, 10)

array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
       0.55555556, 0.66666667, 0.77777778, 0.88888889, 1.        ])

In [42]:
np.linspace(0, 0.1, 10)

array([0.        , 0.01111111, 0.02222222, 0.03333333, 0.04444444,
       0.05555556, 0.06666667, 0.07777778, 0.08888889, 0.1       ])

In [39]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.ElasticNet(alpha=0.07, max_iter=1000
                                    , random_state=1))
])

param_grid = {
    "est__alpha": np.linspace(0, 0.1, 10),
    "est__l1_ratio": np.linspace(0, 1, 10)
}

gsearch = model_selection.GridSearchCV(pipe, param_grid,
                                cv =5, verbose = True)
gsearch.fit(X_train, y_train)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:477: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimato

/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:477: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_

/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:477: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_

/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:267: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  self._final_estimator.fit(Xt, y, **fit_params)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:477: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  positive)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/l

/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of

/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=Pipeline(memory=None,
     steps=[('poly', PolynomialFeatures(degree=1, include_bias=False, interaction_only=False)), ('scaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', ElasticNet(alpha=0.07, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=1, selection='cyclic', tol=0.0001, warm_start=False))]),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'est__alpha': array([0.     , 0.01111, 0.02222, 0.03333, 0.04444, 0.05556, 0.06667,
       0.07778, 0.08889, 0.1    ]), 'est__l1_ratio': array([0.     , 0.11111, 0.22222, 0.33333, 0.44444, 0.55556, 0.66667,
       0.77778, 0.88889, 1.     ])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=True)

In [40]:
gsearch.best_params_

{'est__alpha': 0.07777777777777778, 'est__l1_ratio': 0.1111111111111111}

In [43]:
gsearch.best_estimator_

Pipeline(memory=None,
     steps=[('poly', PolynomialFeatures(degree=1, include_bias=False, interaction_only=False)), ('scaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', ElasticNet(alpha=0.07777777777777778, copy_X=True, fit_intercept=True,
      l1_ratio=0.1111111111111111, max_iter=1000, normalize=False,
      positive=False, precompute=False, random_state=1, selection='cyclic',
      tol=0.0001, warm_start=False))])

In [45]:
y_train_pred = gsearch.predict(X_train)
y_test_pred = gsearch.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))

training r2 0.8973696343723988
test r2 0.8768660857303611
training rmse: 0.12263878092804865
test rmse: 0.1525484051834084


In [46]:
import pickle

In [47]:
with open("model.pkl", "wb") as f:
    pickle.dump(gsearch, f)

In [48]:
with open("model.pkl", "rb") as f:
    model = pickle.load(f)

In [50]:
%%time 

y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))

training r2 0.8973696343723988
test r2 0.8768660857303611
training rmse: 0.12263878092804865
test rmse: 0.1525484051834084
CPU times: user 46.1 ms, sys: 2.02 ms, total: 48.1 ms
Wall time: 12.4 ms
